### ISSUE TITLE
modular front end

### ISSUE THREAD

**Issue by danieleades** [Hirevo/alexandrie#27]
this is a bit of a design question i guess- is there a reason the front end and back end are in a single project? I suppose this means that you don't have to have an API layer between the two, but it makes them tightly coupled. would it not be preferable for the backend to serve some api (REST/gRPC/whatever) for the frontend to consume? More work to set up, perhaps, but possibly a little more flexible and maintainable in the long run. essentially it would make the frontend swappable

**Comment by danieleades** [comment]
by the way, i'm quite impressed with the design decision to use Rust for the frontend. Not the easiest road you could have chosen, but i'm pleased to see early adopters pushing it in this direction!

**Comment by Hirevo** [comment]
Hello,

The reason why I went with a server-rendered frontend was to have it be lightweight and fast to render for the most users.  
The issue with having the frontend query the API is that it requires more roundtrips between the client and the server (one for the application shell and one or more for the API data), which may be costly and add considerable latency in case of a slow connection.  
The server-rendered strategy delivers the page fully rendered to the client, so it avoids this issue.  

Also, while the built-in frontend made these choices, it is completely implemented behind a feature-gate, meaning that it can be disabled and all of the frontend code goes away.  
So, if someone wants to build a richer frontend for Alexandrie, they are completely free to do that since we expose a programmatic API anyway that we aim to be complete and have all the same features that those of the frontend (tracked in **#4**).  

But its true that having it server-rendered adds more maintenance burden on us, but I don't think it is too much to handle.  
Another thing is that we are also a little more limited in how dynamic we can make each page. Here, we are incited to use multiple pages, instead of having one, highly-dynamic page.  
(but we can also sort-of emulate dynamic pages in a few cases using flash cookies, like in the account management page.)

**TL;DR:**  
Basically, the server-rendered frontend was a choice to have it be lightweight for clients.  
We offer a programmatic API anyway to still allow alternative frontends and won't be having exclusive features not available in that API.  
And the frontend can be freely disabled to not pay any of its costs (not even its compile-time), so I figured it was OK.

**Comment by Hirevo** [comment]
Also, an interesting property of server-rendered pages is that it doesn't require any JavaScript to load (we also don't use any JavaScript for responsive layout purposes, it is all pure CSS).  
So, if someone with JavaScript disabled wishes to see a crate's page in Alexandrie, he will be able to access it without problems or limitations.

**Comment by danieleades** [comment]
thanks for taking the time to explain the design decision!

I would assume that generally the frontend and backend would be running on the same host in most deployments, so i'm not sure there'd be a big impact in terms of latency, and i'd probably pay that price just for the separation of concerns, but I totally understand where you're coming from.
and yeah, since it's behind a feature flag, i guess you can just pretend it doesn't exist if you wish.

Thanks again! This is a brilliant project, best of luck!
